packages/go_router/lib/src/match.dart

```diff
@@ -16,77 +16,281 @@ import 'logging.dart';
 import 'misc/errors.dart';
 import 'path_utils.dart';
 import 'route.dart';
+import 'state.dart';
 
-/// An matched result by matching a [RouteBase] against a location.
+/// The function signature for [RouteMatchList.visitRouteMatches]
 ///
-/// This is typically created by calling [RouteMatch.match].
-@immutable
-class RouteMatch {
-  /// Constructor for [RouteMatch].
-  const RouteMatch({
-    required this.route,
-    required this.matchedLocation,
-    required this.pageKey,
-  });
+/// Return false to stop the walk.
+typedef RouteMatchVisitor = bool Function(RouteMatchBase);
+
+/// The base class for various route matches.
+abstract class RouteMatchBase with Diagnosticable {
+  /// An abstract route match base
+  const RouteMatchBase();
+
+  /// The matched route.
+  RouteBase get route;
+
+  /// The page key.
+  ValueKey<String> get pageKey;
+
+  /// The location string that matches the [route].
+  ///
+  /// for example:
+  ///
+  /// uri = '/family/f2/person/p2'
+  /// route = GoRoute('/family/:id')
+  ///
+  /// matchedLocation = '/family/f2'
+  String get matchedLocation;
+
+  /// Gets the state that represent this route match.
+  GoRouterState buildState(
+      RouteConfiguration configuration, RouteMatchList matches);
 
   /// Generate a [RouteMatch] object by matching the `route` with
   /// `remainingLocation`.
   ///
   /// The extracted path parameters, as the result of the matching, are stored
   /// into `pathParameters`.
-  static RouteMatch? match({
+  static List<RouteMatchBase> match({
```

Comment:
This is the most complex part in this pr, previously `match` just produce a RouteMatch for the input route.

After this change, this `match` will recursively match the subroute in `route` and return the entire route matches if it finds the full match.